fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +716,207 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+        getAllPredicates(root, predicates);
+        return predicates;
+    }
+
+    private static void getAllPredicates(OptExpression root, List<ScalarOperator> predicates) {
+        Operator operator = root.getOp();
+        if (operator.getPredicate() != null) {
+            predicates.add(root.getOp().getPredicate());
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            if (joinOperator.getOnPredicate() != null) {
+                predicates.add(joinOperator.getOnPredicate());
+            }
+        }
+        for (OptExpression child : root.getInputs()) {
+            getAllPredicates(child, predicates);
+        }
+    }
+
+    // split predicate into three parts: equal columns predicates, range predicates, and residual predicates
+    public static PredicateSplit splitPredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return PredicateSplit.of(null, null, null);
+        }
+        List<ScalarOperator> predicateConjuncts = Utils.extractConjuncts(predicate);
+        List<ScalarOperator> columnEqualityPredicates = Lists.newArrayList();
+        List<ScalarOperator> rangePredicates = Lists.newArrayList();
+        List<ScalarOperator> residualPredicates = Lists.newArrayList();
+        for (ScalarOperator scalarOperator : predicateConjuncts) {
+            if (scalarOperator instanceof BinaryPredicateOperator) {
+                BinaryPredicateOperator binary = (BinaryPredicateOperator) scalarOperator;
+                ScalarOperator leftChild = scalarOperator.getChild(0);
+                ScalarOperator rightChild = scalarOperator.getChild(1);
+                if (binary.getBinaryType().isEqual()) {
+                    if (leftChild.isColumnRef() && rightChild.isColumnRef()) {
+                        columnEqualityPredicates.add(scalarOperator);
+                    } else if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                } else if (binary.getBinaryType().isRange()) {
+                    if (leftChild.isColumnRef() && rightChild.isConstantRef()) {
+                        rangePredicates.add(scalarOperator);
+                    } else {
+                        residualPredicates.add(scalarOperator);
+                    }
+                }
+            } else {
+                residualPredicates.add(scalarOperator);
+            }
+        }
+        return PredicateSplit.of(Utils.compoundAnd(columnEqualityPredicates), Utils.compoundAnd(rangePredicates),
+                Utils.compoundAnd(residualPredicates));
+    }
+
+    public static ScalarOperator canonizePredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return null;
+        }
```

Comment:
if we move the check to ScalarOperatorRewriter, we need the method?

Reply:
you mean I should use ScalarOperatorVisitor here?

Reply:
I think we should add this method for encapsulation